[hipblaslt][CMS] Add CMS for BBS 160x256x64 TN DTL

shbae opened the pull request:
## Motivation

Add CMS for BBS 160x256x64 TN tile, which improve performance with the tile by CMS.

## Technical Details

Applied CMS optimization for 160x256x64 TN tile, and the new kernel improves the performance as below:
* ~~**main loop efficiency**: improved from 86% efficiency to 96% efficiency.~~
* ~~**tensilelite yaml**: **~5.1%** speedup compared to non-CMS version.~~
* ~~**hipblaslt-bench**: **~4.2%** speedup compared to hipblaslt without adding the CMS kernel for 160x256x64 TN.~~

Updated CMS optimization performance result:
* **main loop efficiency**: improved from 86% efficiency to **94.8%** efficiency.
* **tensilelite yaml**: **~4.7%** speedup compared to non-CMS version.
* **hipblaslt-bench**: **~4.3%** speedup compared to hipblaslt without adding the CMS kernel for 160x256x64 TN.

## Test Result
* Tested through Tensilelite yaml file and all tests PASSED:
```
        - ProblemSizes:
          - Exact: [2560, 4096, 1, 8192]  # [160x16, 256x16, 1, 64x128]
          - Range: [[160], [256], [1], [64, 64, 256]]    # - Range: [[MT0], [MT1], [1], [DU, DU, 4*DU]]
          - Range: [[160], [256], [1], [1,1,64]]         # - Range: [[MT0], [MT1], [1], [1,1,DU]]
          - Range: [[160], [256], [1], [32, 64, 256]]    # - Range: [[MT0], [MT1], [1], [DU/2, DU, 4*DU]]
          - Range: [[5120], [8192], [1], [64, 64, 256]]  # - Range: [[32*MT0], [32*MT1], [1], [DU, DU, 4*DU]]
```
* `hipblaslt-test` with `--gtest_filter="*_TN_*" and all tests PASSED.
```
[----------] Global test environment tear-down
[==========] 5714 tests from 2 test suites ran. (1053435 ms total)
[  PASSED  ] 5714 tests.
hipBLASLt version: 100200
hipBLASLt git version: 1eb7acfc2d-dirty
command line: ./build/release/clients/hipblaslt-test --gtest_filter=*_TN_*
```
With the updated CMS optimization:
```
[----------] Global test environment tear-down
[==========] 5714 tests from 2 test suites ran. (1070277 ms total)
[  PASSED  ] 5714 tests.
hipBLASLt version: 100200
hipBLASLt git version: f5de00f2cf
command line: ./build/release/clients/hipblaslt-test --gtest_filter=*_TN_*
```
## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+        }
+        # note: syncCode needs to be
+        syncCode = [SWaitCnt(dscnt=13, vlcnt=-1, vscnt=-1, comment="Wait for prior LRA0/LRB0 before starting main loop"),
+                    SWaitCnt(dscnt=13, vlcnt=-1, vscnt=-1, comment="Wait for LRA0 to complete"),
```
It should be  SWaitCnt(dscnt=0) if you wait for LRA0 to complete.

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+                    SBarrier(comment="for GRA start"),
+                    SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LRB0 to complete"),
+                    SBarrier(comment="for GRB / LRA1 start"),
+                    SWaitCnt(dscnt=-1, vlcnt=13, vscnt=-1, comment="Wait for GRA to complete"),
```
This wait (SWaitCnt(dscnt=-1, vlcnt=13) should be before the Sbarrier (SBarrier(comment="for GRB / LRA1 start")).
As you say, we wait for GRB / LRA1 start

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+            'LCC'    : [[79, 79]],
+        }
+        # note: syncCode needs to be
+        syncCode = [SWaitCnt(dscnt=13, vlcnt=-1, vscnt=-1, comment="Wait for prior LRA0/LRB0 before starting main loop"),
```
SWaitCnt(dscnt=13) is too high. We need to make sure some of these 13 reads are done before doing MFMAs.

sebvince replied:
We are not waiting for prior LRA0/LRB0 but LRA1/LRB1

shbae commented:
> Thanks for looking at this tile size ! Found some potential race conditions. Please double check the perf once fixed.

Thanks for reviewing my PR and the offline discussion @sebvince! I've fixed this PR based on your suggestions and other offline discussions.  And the performance change is pretty small.  Please refer to the main description for the new performance summary.

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+                    SBarrier(comment=""),
+                    SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LRB0 to complete to start GRB"),
+                    SBarrier(comment=""),
+                    SWaitCnt(dscnt=-1, vlcnt=(13 + 8), vscnt=-1, comment="Wait for GRA to complete to start LRA1"),
```
Count should be 14 instead of 21.

Iteration N :
 - 5 GRA
 - 8 GRB
Iteration N+1:
- 5 GRA
- 1 GRB
- WaitCnt -> We want previous GRA to be done. That's  8 previous GRB + 5 current GRA + 1 current GRB = 14
- ...

shbae replied:
Oh, good catch! Thanks for catching it.

shbae replied:
done

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+                    SWaitCnt(dscnt=5, vlcnt=-1, vscnt=-1, comment="Wait for prior LRA1/LRB1 for the remaining main loop"),
+                    SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LRA0 to complete to start GRA"),
+                    SBarrier(comment=""),
+                    SWaitCnt(dscnt=0, vlcnt=-1, vscnt=-1, comment="Wait for LRB0 to complete to start GRB"),
```
Ok we wait for LRB0 to finish before starting GRB. However you also need to wait on the first LRB0 before MFMA index 40.
Starting from index 40, MFMA instructions will use VPGRs from LRB0.
Either move up a bit this SWaitCnt(dscnt=0) to 39 or add an intermediate SWaitCnt at 39 to guarantee LBR0 used from 40 to 42 are done.

shbae replied:
Well, you right.  But the `SWaitCnt(dscnt=0)` is already scheduled at 38, since GRB starts at 40 anyway. :)

sebvince replied:
Oh right sorry I guess I was thinking this wait was the one at 42. All good then !

sebvince commented on projects/hipblaslt/tensilelite/Tensile/Components/CustomSchedule.py:
```diff
@@ -577,9 +577,57 @@ def _get_schedule_256x256x64_16bit(kernel, useLDSTr, TLDS):
+        }
+        # note: syncCode needs to be
+        syncCode = [SWaitCnt(dscnt=7, vlcnt=-1, vscnt=-1, comment="Wait for necessary prior LRA1/LRB1 before starting main loop"),
+                    SWaitCnt(dscnt=5, vlcnt=-1, vscnt=-1, comment="Wait for prior LRA1/LRB1 for the remaining main loop"),
```
It's good but this is done a bit too late.
Your 1st SWaitCnt(dscnt=7) guarantees you have all previous LRA1 done and 1 LBRB1. Given that you have 5 LRA1, it means you can safely perform 5 MFMAs (1 value of B x 5 values of A).
Starting from index 5, MFMA instructions will start using 2nd value of LBRB1 (vgprValuB_X0_I0[4:7] in asm). However your wait is at index 7 .
So you need a wait at index 4, not 6.

shbae replied:
Yeah, you right. I thought 1 more LRB1 should be done before index 5 (as same as default code doesn't check it there), so I didn't check it there as well. But your suggestion checks it explicitly. I'll update and let you know.

shbae replied:
Done.

shbae commented:
@sebvince I've updated the PR w.r.t. your comments and also rebased this PR on top of the HEAD of `hipblaslt_common_cms_dev`.  The performance improvement is still same as **4.3%** speedup.

shbae commented:
Confirmed that all tests PASSED locally.
```
[----------] Global test environment tear-down
[==========] 5714 tests from 2 test suites ran. (1061207 ms total)
[  PASSED  ] 5714 tests.
hipBLASLt version: 100200
hipBLASLt git version: 27f769869e-dirty
command line: ./build/release/clients/hipblaslt-test --gtest_filter=*_TN_*
```

sebvince reviewed (approved):
LGTM

shbae commented:
@sebvince thank you again for reviewing this PR thoroughly!  I've already rebased off the latest merged PR and the `math-ci preliminary` just PASSED now.  So, I'm merging this branch into `hipblaslt_common_cms_dev` branch.
